### PR TITLE
Fixes #3307: Pushing yet-unseen tombstoned doc to Sync Gateway returns error in XATTR mode

### LINKED ISSUE #3307: Pushing yet-unseen tombstoned doc to Sync Gateway returns error in XATTR mode
## Steps to repro

1. Run the litecore tests via `./build_cmake/macos/C/tests/C4Tests -r list '[.SyncServer]'` (see https://github.com/couchbase/sync_gateway/issues/3286 for more detaisl)
1. Run Sync Gateway against the feature branch from the https://github.com/couchbase/sync_gateway/pull/3306 PR with [this config](https://gist.github.com/tleyden/194ce01acb58e4f0d95589d28b86e057#file-config-json-L50-L67)
1. The  `Push & Pull Deletion` test fails with [test logs](https://gist.github.com/tleyden/354ee8d892dd5352b17c0381648e0d5f#file-gistfile1-txt-L1379) and [sg logs](https://gist.github.com/tleyden/101ab42fb4cef66fd310877d8aaf5a78) 
    - writes a rev
    - writes a tombstone rev for that doc
    - does a one-shot push  <-- test fails here 
    - deletes local db
    - does a  pull
    - verifies the local db has rev2 and it's deleted

## Steps to fix

- [x] Create unit test that uses blip
- [ ] Create unit test that uses REST api to verify if this bug only affects blipsync
- [x] Fix
- [x] Rerun LiteCore suite

### PR DESCRIPTION
**Pull request by tleyden**
Fixes #3307 

Integration test:

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/399/console (XATTR)

### CODE REVIEW

File: base/bucket_gocb.go [hunk 1]
```diff
@@ -1101,12 +1101,22 @@ func (bucket CouchbaseBucketGoCB) UpdateXattr(k string, xattrKey string, exp uin
 	// This is the only use case for macro expansion today - if more cases turn up, should change the sg-bucket API to handle this more generically.
 	xattrCasProperty := fmt.Sprintf("%s.cas", xattrKey)
 	worker := func() (shouldRetry bool, err error, value interface{}) {
-		// If the body doesn't exist, we need to set the AccessDelete flag to mutate the xattr.  If the body exists (and we're trying to delete it), revert
-		// to SubdocDocFlagNone
-		mutateFlag := gocb.SubdocDocFlagAccessDeleted
+
+		var mutateFlag gocb.SubdocDocFlag
 		if deleteBody {
+			// If the body exists (and we're trying to delete it), use SubdocDocFlagNone
```
Comment (adamcfraser) [inline base/bucket_gocb.go]: Better stated as "Since the body exists, we don't need to set a SubdocDocFlag"

File: base/bucket_gocb.go [hunk 2]
```diff
@@ -1101,12 +1101,22 @@ func (bucket CouchbaseBucketGoCB) UpdateXattr(k string, xattrKey string, exp uin
...
 			mutateFlag = gocb.SubdocDocFlagNone
+		} else {
+			if cas == 0 {
+				// If the cas is 0, that means the document doesn't exist, in which case we are trying to set
+				// Xattrs on a non-existent document, so we must set the SubdocDocFlagMkDoc flag
+				mutateFlag = gocb.SubdocDocFlagMkDoc
+			} else {
+				// If the body doesn't exist, we need to set the AccessDelete flag to mutate the xattr.
```
Comment (adamcfraser) [inline base/bucket_gocb.go]: More correct to say "If the body *may* not exist, we need to set SubdocDocFlagAccessDeleted"

File: base/bucket_gocb.go [hunk 3]
```diff
@@ -1101,12 +1101,22 @@ func (bucket CouchbaseBucketGoCB) UpdateXattr(k string, xattrKey string, exp uin
...
-		// to SubdocDocFlagNone
-		mutateFlag := gocb.SubdocDocFlagAccessDeleted
+
+		var mutateFlag gocb.SubdocDocFlag
 		if deleteBody {
+			// If the body exists (and we're trying to delete it), use SubdocDocFlagNone
 			mutateFlag = gocb.SubdocDocFlagNone
+		} else {
+			if cas == 0 {
+				// If the cas is 0, that means the document doesn't exist, in which case we are trying to set
```
Comment (adamcfraser) [inline base/bucket_gocb.go]: Can probably simplify to "If the doc doesn't exist, set SubdocDocFlagMkDoc to allow us to write the xattr"

**Review by adamcfraser — APPROVED** [review]
Looks good - some minor editing suggestions on the comments for simplicity/clarify.

**Comment by tleyden** [comment]
Awesome, thanks.  I updated the wording on all of those comments.

**Comment by tleyden** [comment]
The integration test suite passed.

**Pull request merged**